[Model Monitoring] Fix ML-11518 - Memory leak in Kafka stream pods

alxtkr77 opened the pull request:
### 📝 Description

Fixes ML-11518 - Memory leak in Kafka stream pods caused by orphaned ParquetTarget buffers during consumer group rebalancing.

**Root Cause**: During Kafka rebalancing (~10.64 times/hour), ParquetTarget buffers (~20MB each) were orphaned because the drain callback destroyed the async flow before flushing buffered events. This resulted in 3,183 MB leak over 12.48 hours in production.

---

### 🛠️ Changes Made

- Added `_flush_all_batching_steps()` helper function to walk the graph and flush all batching steps
- Modified `drain_callback()` to call flush BEFORE terminating the async flow
- Added graceful error handling to avoid blocking rebalancing
- Created integration tests using real ParquetStoreyTarget and storey flows

---

### ✅ Checklist

- [x] I updated the documentation (test docstrings)
- [x] I have tested the changes in this PR

---

### 🧪 Testing

**Integration Tests** (tests/serving/test_drain_callback.py):
- Uses **real** ParquetStoreyTarget with actual storey async flows
- Verified: event buffering → flush → Parquet file I/O → cleanup
- Edge cases: empty batches, error handling, callback registration
- **Validation**: Tests FAIL without fix (50 events orphaned), PASS with fix (0 events, files written to disk)

**Evidence from Production (vmdev209)**:
- Memory leak: 3,763 MB vs 580 MB baseline = 3,183 MB over 12.48 hours
- Rebalancing frequency: 10.64 times/hour
- Calculated orphaned batches: 159 (3,183 MB / 20 MB per batch)
- Expected orphaned batches: 133 (10.64/hr × 12.48hr)
- Variance: 20% (acceptable given measurement windows)

---

### 🔗 References

- Ticket link: ML-11518
- Design docs links: N/A
- External links: N/A

---

### 🚨 Breaking Changes?

- [x] No

---

### 🔍️ Additional Notes

**Mechanism**: The fix ensures that when Kafka triggers a consumer group rebalance:
1. Drain callback is invoked
2. All batching steps (ParquetTarget) are explicitly flushed
3. Buffered events are written to storage
4. THEN the async flow is terminated cleanly
5. No orphaned buffers remain in memory

**Next Step**: Deploy to vmdev209 to verify the fix resolves the production memory leak.

Commit:
[Model Monitoring] Fix ML-11518 - Memory leak in Kafka stream pods

Add drain callback flush to prevent ParquetTarget buffer orphaning during rebalancing.

assaf758 commented on mlrun/serving/server.py:
```diff
@@ -526,6 +526,34 @@ def add_system_steps_to_graph(
+
+            # Recursively flush child steps
+            if hasattr(step, "_outlets"):
+                await _flush_all_batching_steps(step, logger)
```
what happens if the DAG is:
Buffer1 => step-X => Buffer2
Will the recursive "find" the event in Buffer2, or it will still be "in-flight" at step-X in which case it will leak?

assaf758 commented on mlrun/serving/server.py:
```diff
@@ -892,6 +920,16 @@ async def termination_callback():
+                    server.graph._async_flow, context.logger
+                )
+                context.logger.info("Batched data flushed successfully")
+
```
I would explain here why `wait_for_completion` is not enough

assaf758 reviewed (commented):
Nice!
1. Cyclic-graph (mlrun-10753) is coming in 1.11, need to make sure that this recursive solution can handle it.
2. Seems like the fix is (and testing) will be more natural in storey - what do you think?

alxtkr77 commented:
Thanks for the review! I've addressed the feedback:

## 1. Cyclic graphs (ML-10753) ✓
Added cycle detection using a `visited` set to handle cyclic graphs safely.

## 2. Why wait_for_completion() is not enough ✓
Added detailed comment explaining the distinction between in-flight events (handled by wait_for_completion) vs buffered events (require explicit flush).

## 3. Buffer1 => step-X => Buffer2 scenario

Good question about event propagation! Let me trace through:

**Execution flow:**
1. Visit Buffer1: `await Buffer1._emit_all()` - emit buffered events
2. Recurse to step-X (not batching, skip)  
3. Recurse to Buffer2: `await Buffer2._emit_all()` - might be empty if events in-flight

**Key question**: Does `await step._emit_all()` wait for downstream propagation?

The full drain sequence is:
1. `_flush_all_batching_steps()` - recursively flush all buffers
2. `wait_for_completion()` - wait for ALL in-flight events

Even if Buffer2 is empty during flush, events in-flight at step-X will complete during wait_for_completion(). These events should be ACK'd properly as part of the in-flight completion.

Will investigate further.

## 4. Storey vs MLRun fix location

Drain callback is Nuclio-specific (Storey doesn't know about Nuclio rebalancing).

Commit:
[Model Monitoring] Address PR #8942 review feedback

- Add cycle detection to _flush_all_batching_steps() for ML-10753 cyclic graph support
- Add detailed comment explaining why wait_for_completion() is not sufficient
- Add test for cyclic graph handling to prevent regression

ML-11518

Commit:
[Model Monitoring] Implement multi-pass flushing for cyclic graphs

- Define MAX_FLUSH_ITERATIONS constant (5 iterations)
- Split flush logic into single-pass and multi-pass functions
- Multi-pass flushing handles events that re-enter nodes via cycles
- Each iteration resets visited set and flushes until stable
- Logging for diagnostics and warning if not stabilized
- Adds test for cyclic graphs with event re-entry

This ensures ML-10753 cyclic graphs are properly handled without
needing the step revisit counter. Events that propagate back to
already-visited nodes via cycles will be flushed in subsequent passes.

ML-11518

assaf758 commented on mlrun/serving/server.py:
```diff
@@ -526,6 +530,96 @@ def add_system_steps_to_graph(
+        for step in flow._outlets:
+            # Skip if already visited (handles cyclic graphs)
+            if id(step) in visited:
+                continue
```
As I understand you are doing multiple iterations saving `visited` across iterations, and then skipping an event on step revisit... But why? The whole point is to allow cyclic flows down the road, where you do have revisits..

alxtkr77 replied:
The visited set is reset on each iteration, so cyclic events are flushed. The single-pass function uses visited only to prevent infinite loops within one pass, then the outer loop resets it and tries again.

assaf758 replied:
I see.
So just for clarity, maybe we can remove `visited` as parameter (anyway I don't see it is used in the calling function)

alxtkr77 replied:
It is required and used. It is got reset at line 596

assaf758 commented on mlrun/serving/server.py:
```diff
@@ -526,6 +530,96 @@ def add_system_steps_to_graph(
+
+    except Exception as e:
+        logger.warning(f"Error flushing batching steps during drain: {e}")
+        # Don't fail the drain - better to lose some data than block rebalancing
```
I'm not sure about this statement - especially as I'm afraid explicit-ack logic will be violated by this.

alxtkr77 replied:
I understand you suggest failing the drain - raise exception. Correct?

assaf758 replied:
actually I think it's ok - there will be leak of the non-flush, but the new worker receive the non-acked events.

assaf758 commented:
> Drain callback is Nuclio-specific (Storey doesn't know about Nuclio rebalancing).

That's correct, but drain_callback calls functions in storey to do the actual work, so I think they can be tested there with less boilerplate and mocks.

assaf758 reviewed (commented):
> That's correct, but drain_callback calls functions in storey to do the actual work, so I think they can be tested there with less boilerplate and mocks.

Please also consider my comment. Adding @gtopper as well to the review - what you think?

Answered inline.

gtopper commented on mlrun/serving/server.py:
```diff
@@ -892,6 +986,20 @@ async def termination_callback():
+
+            # Force flush all batching steps (e.g., ParquetTarget) before terminating.
+            # wait_for_completion() below waits for in-flight events to complete processing,
+            # but it does NOT force buffered batches to flush. Batching steps like ParquetTarget
```
This isn't true though. In https://github.com/mlrun/storey/blob/7f3f3c8768585ec75bb458b798b58916b9c735c6/storey/flow.py#L1136, a termination event triggers a call to `_emit_all()`.

gtopper reviewed (changes requested):
Let's discuss this. If the drain callback doesn't await serving graph termination and returns immediately, that's a bug – likely a regression – and we need to understand what happened. Storey already has graceful termination built into it (see comment below). If there's an issue with that mechanism, we're able to change the code there.

gtopper commented on tests/serving/test_drain_callback.py:
```diff
@@ -0,0 +1,409 @@
+# limitations under the License.
+
+"""
+Integration tests for drain callback behavior during Kafka rebalancing.
```
It's a unit test, though.

gtopper commented on tests/serving/test_drain_callback.py:
```diff
@@ -0,0 +1,409 @@
+# Copyright 2024 Iguazio
```
```suggestion
# Copyright 2025 Iguazio
```

gtopper commented on tests/serving/test_drain_callback.py:
```diff
@@ -0,0 +1,409 @@
+        finally:
+            # Ensure flow is terminated even if test fails
+            # This prevents orphaned background tasks and file handles
+            await controller.terminate()
```
`controller.terminate()` and `controller.await_termination()` need to be called in order to terminate the graph. They do not belong in test cleanup – calling these functions should flush all events, and they should therefore be called before the relevant assertions.
